File: tdf-hq-ui/src/api/merchReputation.ts
import { get, post, put } from './client';

export type MerchReputationState = 'new_store' | 'unrated' | 'published';
export type MerchConfidence = 'new' | 'limited' | 'moderate' | 'strong';

export interface MerchDimensionAggregate {
  code: string;
  average: number | null;
  count: number;
  distribution?: Record<string, number>;
}

export interface MerchReputationSummary {
  subjectKind?: 'store' | 'product';
  id?: string;
  storeId?: string;
  productId?: string;
  storeName?: string;
  productName?: string;
  name?: string;
  slug?: string;
  commercialReputation?: boolean;
  reputationLabel?: string;
  state: MerchReputationState;
  rating: number | null;
  verifiedReviewCount?: number;
  verifiedPurchaseReviewCount?: number;
  historicalReviewCount?: number;
  confidence?: MerchConfidence;
  formulaVersion?: string | null;
  dimensions?: MerchDimensionAggregate[];
  objectiveSignals?: {
    identityVerified?: boolean;
    platformMemberSince?: string;
  };
  badges?: { code: string; earnedAt: string; validUntil: string }[];
}

export interface MerchReviewEligibility {
  orderId: string;
  storeId: string;
  orderState?: string;
  fulfillmentState?: string;
  storeReview: {
    eligible: boolean;
    state: 'available' | 'edit_available' | 'period_expired';
    reviewId: string | null;
    currentRevision: number;
    deadline: string | null;
  };
  productLines: {
    lineId: string;
    productId: string;
    productName: string;
    fulfillmentState: string;
    eligible: boolean;
    state: 'available' | 'edit_available' | 'period_expired';
    reviewId: string | null;
    currentRevision: number;
  }[];
}

export interface MerchReviewSubmit {
  overallRating: number;
  issueOccurred: boolean;
  comment?: string;
  dimensions: Record<string, number>;
  images?: { mediaAssetId: string; altText: string }[];
  expectedRevision: number;
}

export interface MerchReviewPublic {
  id: string;
  kind: 'store' | 'product';
  rating: number;
  comment: string | null;
  status: 'published' | 'limited';
  verifiedPurchase: true;
  badge: 'Compra verificada';
  author: { name: string; avatarUrl?: string | null };
  dimensions: Record<string, number>;
  images: { assetId: string; url: string; altText: string; position: number }[];
  sellerResponse?: { id: string; body: string; updatedAt: string; status: string } | null;
  createdAt: string;
  updatedAt: string;
}

export interface MerchReputationPriorities {
  subjectKind: 'store' | 'product';
  revision: number;
  affectsPublicScore: false;
  orderedDimensions: {
    code: string;
    nameEs: string;
    nameEn: string;
    definitionEs: string;
    definitionEn: string;
  }[];
}

export interface MerchNotificationPreferences {
  reviewInvitation: boolean;
  reviewReminder: boolean;
  sellerResponseNotification: boolean;
  moderationChange: boolean;
  evidenceRequest: boolean;
  appealResult: boolean;
  badgeChange: boolean;
}

const idempotency = (key?: string) => ({
  headers: { 'Idempotency-Key': key ?? crypto.randomUUID() },
});

export const MerchReputation = {
  artistStores: (artistPartyId: number) =>
    get<MerchReputationSummary[]>('/merch/artists/' + artistPartyId + '/stores'),
  store: (storeId: string) =>
    get<MerchReputationSummary>('/merch/stores/' + encodeURIComponent(storeId) + '/reputation'),
  product: (productId: string) =>
    get<MerchReputationSummary>('/merch/products/' + encodeURIComponent(productId) + '/reputation'),
  reviews: (kind: 'store' | 'product', subjectId: string, cursor?: string) => {
    const query = cursor ? '?cursor=' + encodeURIComponent(cursor) : '';
    const plural = kind === 'store' ? 'stores' : 'products';
    return get<{ items: MerchReviewPublic[]; nextCursor?: string | null }>(
      '/merch/' + plural + '/' + encodeURIComponent(subjectId) + '/reviews' + query,
    );
  },
  formula: (locale: 'es' | 'en' = 'es') =>
    get<Record<string, unknown>>('/merch/reputation/formula?locale=' + locale),
  eligibility: (orderId: string) =>
    get<MerchReviewEligibility>('/merch/orders/' + encodeURIComponent(orderId) + '/reviews/eligibility'),
  submitStore: (orderId: string, body: MerchReviewSubmit, key?: string) =>
    put<Record<string, unknown>>(
      '/merch/orders/' + encodeURIComponent(orderId) + '/store-review',
      body,
      idempotency(key),
    ),
  submitProduct: (orderId: string, lineId: string, body: MerchReviewSubmit, key?: string) =>
    put<Record<string, unknown>>(
      '/merch/orders/' + encodeURIComponent(orderId) + '/lines/' + encodeURIComponent(lineId) + '/product-review',
      body,
      idempotency(key),
    ),
  respond: (reviewId: string, responseBody: string, responseExpectedRevision: number, key?: string) =>
    put<Record<string, unknown>>(
      '/merch/reviews/' + encodeURIComponent(reviewId) + '/response',
      { responseBody, responseExpectedRevision },
      idempotency(key),
    ),
  report: (
    body: {
      reportTargetType: 'review' | 'seller_response';
      reportTargetId: string;
      reportReason: string;
      reportDetails?: string;
      authorizedEvidence?: unknown[];
    },
    key?: string,
  ) => post<Record<string, unknown>>('/merch/reputation/reports', body, idempotency(key)),
  appeal: (decisionId: string, appealGrounds: string, key?: string) =>
    post<Record<string, unknown>>(
      '/merch/reputation/decisions/' + encodeURIComponent(decisionId) + '/appeal',
      { appealGrounds },
      idempotency(key),
    ),
  sellerStore: (storeId: string) =>
    get<Record<string, unknown>>('/merch/seller/stores/' + encodeURIComponent(storeId) + '/reputation'),
  moderationCases: (state?: string) =>
    get<Record<string, unknown>[]>(
      '/merch/admin/reputation/cases' + (state ? '?state=' + encodeURIComponent(state) : ''),
    ),
  transitionCase: (
    caseId: string,
    moderationAction: 'triage' | 'request_evidence' | 'provisionally_hide' | 'resume_review',
    workflowRationale: string,
    key?: string,
  ) => post<Record<string, unknown>>(
    '/merch/admin/reputation/cases/' + encodeURIComponent(caseId) + '/workflow',
    { moderationAction, workflowRationale, workflowEvidence: { reviewedInAdminPanel: true } },
    idempotency(key),
  ),
  decide: (caseId: string, body: Record<string, unknown>, key?: string) =>
    post<Record<string, unknown>>(
      '/merch/admin/reputation/cases/' + encodeURIComponent(caseId) + '/decision',
      body,
      idempotency(key),
    ),
  resolveAppeal: (
    appealId: string,
    appealOutcome: 'upheld' | 'reversed',
    appealRationale: string,
    key?: string,
  ) => post<Record<string, unknown>>(
    '/merch/admin/reputation/appeals/' + encodeURIComponent(appealId) + '/decision',
    { appealOutcome, appealRationale, appealEvidence: { reviewedInAdminPanel: true } },
    idempotency(key),
  ),
  priorities: (subjectKind: 'store' | 'product') =>
    get<MerchReputationPriorities>('/merch/reputation/preferences/' + subjectKind),
  savePriorities: (
    subjectKind: 'store' | 'product',
    orderedDimensionCodes: string[],
    priorityExpectedRevision: number,
    key?: string,
  ) => put<MerchReputationPriorities>(
    '/merch/reputation/preferences/' + subjectKind,
    { orderedDimensionCodes, priorityExpectedRevision },
    idempotency(key),
  ),
  suggestCategory: (
    suggestionSubjectKind: 'store' | 'product',
    suggestionLabel: string,
    suggestionDefinition: string,
    key?: string,
  ) => post<Record<string, unknown>>(
    '/merch/reputation/category-suggestions',
    { suggestionSubjectKind, suggestionLabel, suggestionDefinition },
    idempotency(key),
  ),
  notificationPreferences: () =>
    get<MerchNotificationPreferences>('/merch/reputation/notification-preferences'),
  saveNotificationPreferences: (body: MerchNotificationPreferences) =>
    put<MerchNotificationPreferences>('/merch/reputation/notification-preferences', body),
  categorySuggestions: (status?: string) => get<Record<string, unknown>[]>(
    '/merch/admin/reputation/category-suggestions' + (status ? '?status=' + encodeURIComponent(status) : ''),
  ),
  decideCategorySuggestion: (
    suggestionId: string,
    body: Record<string, unknown>,
    key?: string,
  ) => post<Record<string, unknown>>(
    '/merch/admin/reputation/category-suggestions/' + encodeURIComponent(suggestionId) + '/decision',
    body,
    idempotency(key),
  ),
};

File: tdf-hq/src/TDF/API/MerchReputation.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeOperators #-}

module TDF.API.MerchReputation where

import Data.Aeson (FromJSON, ToJSON, Value)
import Data.Int (Int64)
import Data.Text (Text)
import Data.UUID (UUID)
import GHC.Generics (Generic)
import Servant

data MerchReviewImageInput = MerchReviewImageInput
  { mediaAssetId :: UUID
  , altText :: Text
  } deriving (Show, Generic)

instance FromJSON MerchReviewImageInput
instance ToJSON MerchReviewImageInput

data MerchReviewSubmitRequest = MerchReviewSubmitRequest
  { overallRating :: Int
  , issueOccurred :: Bool
  , comment :: Maybe Text
  , dimensions :: Value
  , images :: Maybe [MerchReviewImageInput]
  , expectedRevision :: Int
  } deriving (Show, Generic)

instance FromJSON MerchReviewSubmitRequest
instance ToJSON MerchReviewSubmitRequest

data MerchSellerResponseRequest = MerchSellerResponseRequest
  { responseBody :: Text
  , responseExpectedRevision :: Int
  } deriving (Show, Generic)

instance FromJSON MerchSellerResponseRequest
instance ToJSON MerchSellerResponseRequest

data MerchContentReportRequest = MerchContentReportRequest
  { reportTargetType :: Text
  , reportTargetId :: UUID
  , reportReason :: Text
  , reportDetails :: Maybe Text
  , authorizedEvidence :: Maybe [Value]
  } deriving (Show, Generic)

instance FromJSON MerchContentReportRequest
instance ToJSON MerchContentReportRequest

data MerchModerationDecisionRequest = MerchModerationDecisionRequest
  { moderationDecision :: Text
  , moderationReasonCode :: Text
  , moderationRationale :: Text
  , moderationEvidence :: Value
  } deriving (Show, Generic)

instance FromJSON MerchModerationDecisionRequest
instance ToJSON MerchModerationDecisionRequest

data MerchModerationWorkflowRequest = MerchModerationWorkflowRequest
  { moderationAction :: Text
  , workflowRationale :: Text
  , workflowEvidence :: Value
  } deriving (Show, Generic)

instance FromJSON MerchModerationWorkflowRequest
instance ToJSON MerchModerationWorkflowRequest

data MerchAppealDecisionRequest = MerchAppealDecisionRequest
  { appealOutcome :: Text
  , appealRationale :: Text
  , appealEvidence :: Value
  } deriving (Show, Generic)

instance FromJSON MerchAppealDecisionRequest
instance ToJSON MerchAppealDecisionRequest

data MerchAppealRequest = MerchAppealRequest
  { appealGrounds :: Text
  } deriving (Show, Generic)

instance FromJSON MerchAppealRequest
instance ToJSON MerchAppealRequest

data MerchPriorityRequest = MerchPriorityRequest
  { orderedDimensionCodes :: [Text]
  , priorityExpectedRevision :: Int
  } deriving (Show, Generic)

instance FromJSON MerchPriorityRequest
instance ToJSON MerchPriorityRequest

data MerchCategorySuggestionRequest = MerchCategorySuggestionRequest
  { suggestionSubjectKind :: Text
  , suggestionLabel :: Text
  , suggestionDefinition :: Text
  } deriving (Show, Generic)

instance FromJSON MerchCategorySuggestionRequest
instance ToJSON MerchCategorySuggestionRequest

data MerchCategorySuggestionDecisionRequest = MerchCategorySuggestionDecisionRequest
  { suggestionStatus :: Text
  , suggestionMinimumSample :: Maybe Int
  , suggestionBiasTest :: Maybe Value
  , suggestionUtilityTest :: Maybe Value
  , suggestionDecisionReason :: Text
  } deriving (Show, Generic)

instance FromJSON MerchCategorySuggestionDecisionRequest
instance ToJSON MerchCategorySuggestionDecisionRequest

data MerchNotificationPreferenceRequest = MerchNotificationPreferenceRequest
  { reviewInvitation :: Bool
  , reviewReminder :: Bool
  , sellerResponseNotification :: Bool
  , moderationChange :: Bool
  , evidenceRequest :: Bool
  , appealResult :: Bool
  , badgeChange :: Bool
  } deriving (Show, Generic)

instance FromJSON MerchNotificationPreferenceRequest
instance ToJSON MerchNotificationPreferenceRequest

type RequiredMerchIdempotency =
  Header' '[Required, Strict] "Idempotency-Key" Text

type MerchReputationPublicAPI =
       "merch" :> "artists" :> Capture "artistPartyId" Int64 :> "stores"
         :> Get '[JSON] [Value]
  :<|> "merch" :> "stores" :> Capture "storeId" UUID :> "reputation"
         :> Get '[JSON] Value
  :<|> "merch" :> "stores" :> Capture "storeId" UUID :> "reviews"
         :> QueryParam "cursor" UUID :> QueryParam "limit" Int :> Get '[JSON] Value
  :<|> "merch" :> "products" :> Capture "productId" UUID :> "reputation"
         :> Get '[JSON] Value
  :<|> "merch" :> "products" :> Capture "productId" UUID :> "reviews"
         :> QueryParam "cursor" UUID :> QueryParam "limit" Int :> Get '[JSON] Value
  :<|> "merch" :> "reputation" :> "formula"
         :> QueryParam "locale" Text :> Get '[JSON] Value

type MerchReputationProtectedAPI =
       "merch" :> "orders" :> Capture "orderId" UUID :> "reviews" :> "eligibility"
         :> Get '[JSON] Value
  :<|> "merch" :> "orders" :> Capture "orderId" UUID :> "store-review"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchReviewSubmitRequest
         :> Put '[JSON] Value
  :<|> "merch" :> "orders" :> Capture "orderId" UUID :> "lines"
         :> Capture "lineId" UUID :> "product-review"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchReviewSubmitRequest
         :> Put '[JSON] Value
  :<|> "merch" :> "reviews" :> Capture "reviewId" UUID :> "response"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchSellerResponseRequest
         :> Put '[JSON] Value
  :<|> "merch" :> "reputation" :> "reports"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchContentReportRequest
         :> PostCreated '[JSON] Value
  :<|> "merch" :> "reputation" :> "decisions" :> Capture "decisionId" UUID :> "appeal"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchAppealRequest
         :> PostCreated '[JSON] Value
  :<|> "merch" :> "reputation" :> "preferences" :> Capture "subjectKind" Text
         :> Get '[JSON] Value
  :<|> "merch" :> "reputation" :> "preferences" :> Capture "subjectKind" Text
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchPriorityRequest
         :> Put '[JSON] Value
  :<|> "merch" :> "reputation" :> "category-suggestions"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchCategorySuggestionRequest
         :> Post '[JSON] Value
  :<|> "merch" :> "reputation" :> "notification-preferences"
         :> Get '[JSON] Value
  :<|> "merch" :> "reputation" :> "notification-preferences"
         :> ReqBody '[JSON] MerchNotificationPreferenceRequest :> Put '[JSON] Value
  :<|> "merch" :> "seller" :> "stores" :> Capture "storeId" UUID :> "reputation"
         :> Get '[JSON] Value
  :<|> "merch" :> "admin" :> "reputation" :> "cases"
         :> QueryParam "state" Text :> Get '[JSON] [Value]
  :<|> "merch" :> "admin" :> "reputation" :> "category-suggestions"
         :> QueryParam "status" Text :> Get '[JSON] [Value]
  :<|> "merch" :> "admin" :> "reputation" :> "category-suggestions"
         :> Capture "suggestionId" UUID :> "decision" :> RequiredMerchIdempotency
         :> ReqBody '[JSON] MerchCategorySuggestionDecisionRequest :> Post '[JSON] Value
  :<|> "merch" :> "admin" :> "reputation" :> "cases" :> Capture "caseId" UUID :> "workflow"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchModerationWorkflowRequest
         :> Post '[JSON] Value
  :<|> "merch" :> "admin" :> "reputation" :> "cases" :> Capture "caseId" UUID :> "decision"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchModerationDecisionRequest
         :> Post '[JSON] Value
  :<|> "merch" :> "admin" :> "reputation" :> "appeals" :> Capture "appealId" UUID :> "decision"
         :> RequiredMerchIdempotency :> ReqBody '[JSON] MerchAppealDecisionRequest
         :> Post '[JSON] Value
